library/alloc/src/sync.rs

```diff
@@ -2313,26 +2352,23 @@ impl<T: ?Sized, A: Allocator> Arc<T, A> {
     ///
     /// Note that this requires locking the weak ref count.
     fn is_unique(&mut self) -> bool {
-        // lock the weak pointer count if we appear to be the sole weak pointer
-        // holder.
-        //
-        // The acquire label here ensures a happens-before relationship with any
-        // writes to `strong` (in particular in `Weak::upgrade`) prior to decrements
-        // of the `weak` count (via `Weak::drop`, which uses release). If the upgraded
-        // weak ref was never dropped, the CAS here will fail so we do not care to synchronize.
-        if self.inner().weak.compare_exchange(1, usize::MAX, Acquire, Relaxed).is_ok() {
-            // This needs to be an `Acquire` to synchronize with the decrement of the `strong`
-            // counter in `drop` -- the only access that happens when any but the last reference
-            // is being dropped.
-            let unique = self.inner().strong.load(Acquire) == 1;
-
-            // The release write here synchronizes with a read in `downgrade`,
-            // effectively preventing the above read of `strong` from happening
-            // after the write.
-            self.inner().weak.store(1, Release); // release the lock
-            unique
-        } else {
-            false
+        match self.inner().strong.load(Relaxed) {
+            ONE_STRONG => true, // This is the only Arc, and no Weak pointers exist.
+            ONE_STRONG_WITH_WEAK
+                if self
+                    .inner()
+                    .weak
+                    .compare_exchange(NO_STRONG_WITH_WEAK, STRONG_DROPPED, Acquire, Relaxed)
```

Comment:
This line seems to be using incorrect constants.
```suggestion
                    .compare_exchange(ONE_WEAK, NO_WEAK, Acquire, Relaxed)
```
Though, I guess the `weak` has the lock bit set, so maybe it is something else?

Reply:
Oh oops. This should have been `ONE_WEAK` and `WEAK_LOCKED`. They have the same values though, so this doesn't change correctness, it was just very confusing. ^^'